### PR TITLE
LimitMalloc: instrumentation to dump all disallowed stacks

### PR DESCRIPTION
**Pull request by rpoyner-tri**
I've found this useful to capture all the stack traces at once from
unwanted allocations. The resulting data is a bit raw, but can be easily
post-processed to count/sort events, provide IDE-compatible error
messages indicating source lines, etc.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14899)

### CODE REVIEW

**Review by rpoyner-tri — COMMENTED** [review]
FYI @jwnimmer-tri @ggould-tri @sherm1 @seancurtis-tri useful? too crude?
+@ericcousineau-tri for feature review (yep, glutton for punishment ;) )
+(priority: low) 

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14899#-:-MYAnMeS6scqqS2Q8dGP:b-b48531)* status: LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers (waiting on @EricCousineau-TRI)

File: common/test_utilities/limit_malloc.cc [hunk 1]
```diff
@@ -168,6 +170,13 @@ void Monitor::ObserveAllocation() {
 
   if (!failure) { return; }
 
+  // Non-fatal breakpoint action; use with helper script:
+  // tools/dynamic_analysis/dump_limit_malloc_stacks
+  if (std::getenv("DRAKE_LIMIT_MALLOC_NONFATAL_BREAKPOINT")) {
+    ::raise(SIGTRAP);
```
Comment (jwnimmer-tri) [inline common/test_utilities/limit_malloc.cc:176]: nit Be sure this (and the new include) builds on macOS.

**Comment by jwnimmer-tri** [comment]
I think the general idea is fine.  If someone else finds this useful, I think its fine for master.

**Comment by rpoyner-tri** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYAq4KK84A-rcB6FYSd:-MYAq4KK84A-rcB6FYSe:bhl2802) (no related file):*
@drake-jenkins-bot mac-catalina-clang-bazel-experimental-debug please.

---

**Review by EricCousineau-TRI — COMMENTED** [review]
First pass; LGTM for general idea + usefulness!

Requesting concrete reproduction case to limit need for reviewer creativity :grimacing: 

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14899#-:-MYB4od9CmKHbdd3wvdZ:bjo3lyp)* status: 6 unresolved discussions, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers (waiting on @rpoyner-tri)

___
*[common/test_utilities/limit_malloc.cc, line 175 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYB456H3A_daP3zExuM:-MYB456I-TpLSRWUXLwK:b-jmokf0) ([raw file](https://github.com/robotlocomotion/drake/blob/30f067991444dc9364f15c4fba698d08d3a0eaa4/common/test_utilities/limit_malloc.cc#L175)):*
> ```cpp
>   // Non-fatal breakpoint action; use with helper script:
>   // tools/dynamic_analysis/dump_limit_malloc_stacks
>   if (std::getenv("DRAKE_LIMIT_MALLOC_NONFATAL_BREAKPOINT")) {
> ```

nit User contract may not be sufficiently constrained, may admit confusion; concretely, this may trigger if the env var is set, but empty.

Suggestions (assuming appropriate prior null check)
- Explicitly check for `string(env) == "1"` (fully constrained)
- Just check for `!string(env).empty()`
___
*[tools/dynamic_analysis/dump_limit_malloc_stacks, line 1 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYB4sxM7_hoHC_IlnNZ:-MYB4sxM7_hoHC_IlnN_:b-3in6ny) ([raw file](https://github.com/robotlocomotion/drake/blob/30f067991444dc9364f15c4fba698d08d3a0eaa4/tools/dynamic_analysis/dump_limit_malloc_stacks#L1)):*
> ```
> #!/bin/bash
> ```

Can you ensure that there's a concrete way to show this?

(It may easy enough (5-10min?) to write a Python unittest for this script, if you have concrete example?)
___
*[tools/dynamic_analysis/dump_limit_malloc_stacks, line 2 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYB3t0h4f9aMnrjQrrc:-MYB3t0i4PO3jvEKWlGi:b-cxxkr4) ([raw file](https://github.com/robotlocomotion/drake/blob/30f067991444dc9364f15c4fba698d08d3a0eaa4/tools/dynamic_analysis/dump_limit_malloc_stacks#L2)):*
> ```
> #!/bin/bash
> #
> ```

Should have `set -eu -o pipefail`.

If you think it'll need debugging, please also add `-x`.
___
*[tools/dynamic_analysis/dump_limit_malloc_stacks, line 11 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYB4khJ10m6IhNgs2EU:-MYB4khJ10m6IhNgs2EV:b3kv0zu) ([raw file](https://github.com/robotlocomotion/drake/blob/30f067991444dc9364f15c4fba698d08d3a0eaa4/tools/dynamic_analysis/dump_limit_malloc_stacks#L11)):*
> ```
> # Example:
> #
> #  $ bazel build -c dbg //my_thing:my_test
> ```

nit Consider making this a concrete example (even if it doesn't explicitly fail).
___
*[tools/dynamic_analysis/dump_limit_malloc_stacks, line 13 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYB4bvYB2DuuqvMMIXM:-MYB4bvYB2DuuqvMMIXN:b-7q6ss2) ([raw file](https://github.com/robotlocomotion/drake/blob/30f067991444dc9364f15c4fba698d08d3a0eaa4/tools/dynamic_analysis/dump_limit_malloc_stacks#L13)):*
> ```
> #  $ bazel build -c dbg //my_thing:my_test
> #  $ tools/dynamic_analysis/dump_limit_malloc_stacks \
> #       bazel-bin/my_thing/my_test |& tee stacks.out
> ```

BTW TIL about `|&` - woot! I've just been using the more kludgy `2>&1 | ` (or whatever the right ordering is)
___
*[tools/dynamic_analysis/dump_limit_malloc_stacks, line 38 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYB4-QY8BYJIWbiOi0k:-MYB4-QY8BYJIWbiOi0l:b-bhet4f) ([raw file](https://github.com/robotlocomotion/drake/blob/30f067991444dc9364f15c4fba698d08d3a0eaa4/tools/dynamic_analysis/dump_limit_malloc_stacks#L38)):*
> ```
> eof
> 
> DRAKE_LIMIT_MALLOC_NONFATAL_BREAKPOINT=1 \
> ```

nit Consider using `exec` if there's nothing else after this.

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14899)* status: 6 unresolved discussions, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers (waiting on @rpoyner-tri)

___
*[common/test_utilities/limit_malloc.cc, line 175 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYB456H3A_daP3zExuM:-MYBGOmx-iaPqjtkvZas:b-43hd6a) ([raw file](https://github.com/robotlocomotion/drake/blob/30f067991444dc9364f15c4fba698d08d3a0eaa4/common/test_utilities/limit_malloc.cc#L175)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

nit User contract may not be sufficiently constrained, may admit confusion; concretely, this may trigger if the env var is set, but empty.

Suggestions (assuming appropriate prior null check)
- Explicitly check for `string(env) == "1"` (fully constrained)
- Just check for `!string(env).empty()`
</blockquote></details>

I don't think I can trust `std::string` not to allocate, so things will be a bit more primitive...

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14899)* status: 6 unresolved discussions, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers (waiting on @EricCousineau-TRI and @rpoyner-tri)

___
*[tools/dynamic_analysis/dump_limit_malloc_stacks, line 1 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYB4sxM7_hoHC_IlnNZ:-MYBHEAN7UiWLliE-CkT:b-fj707m) ([raw file](https://github.com/robotlocomotion/drake/blob/30f067991444dc9364f15c4fba698d08d3a0eaa4/tools/dynamic_analysis/dump_limit_malloc_stacks#L1)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Can you ensure that there's a concrete way to show this?

(It may easy enough (5-10min?) to write a Python unittest for this script, if you have concrete example?)
</blockquote></details>

Food for thought: limit-malloc is a test-only diagnostic tool. How much recursive testing is enough?

Also, will gdb even work under CI? `ptrace (2)` in the cloud seems fraught at best.
___
*[tools/dynamic_analysis/dump_limit_malloc_stacks, line 2 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYB3t0h4f9aMnrjQrrc:-MYBI6-173-Vbs9gqB4S:bvu5e8h) ([raw file](https://github.com/robotlocomotion/drake/blob/30f067991444dc9364f15c4fba698d08d3a0eaa4/tools/dynamic_analysis/dump_limit_malloc_stacks#L2)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Should have `set -eu -o pipefail`.

If you think it'll need debugging, please also add `-x`.
</blockquote></details>

FWIW, you can always externally kick on debugging of bash scripts. For script `foo`, you can invoke it via `bash -vx foo` and get plenty of diagnostics. If it were invoked by CI, then I'd be in favor of internal `-x`.

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14899)* status: 3 unresolved discussions, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers (waiting on @EricCousineau-TRI and @rpoyner-tri)

___
*[tools/dynamic_analysis/dump_limit_malloc_stacks, line 1 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYB4sxM7_hoHC_IlnNZ:-MYBR90u-sfksZorCwQY:b-x7gsqu) ([raw file](https://github.com/robotlocomotion/drake/blob/30f067991444dc9364f15c4fba698d08d3a0eaa4/tools/dynamic_analysis/dump_limit_malloc_stacks#L1)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

Food for thought: limit-malloc is a test-only diagnostic tool. How much recursive testing is enough?

Also, will gdb even work under CI? `ptrace (2)` in the cloud seems fraught at best.
</blockquote></details>

I'm inclined to punt on another level of automated test at this point.
___
*[tools/dynamic_analysis/dump_limit_malloc_stacks, line 2 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYB3t0h4f9aMnrjQrrc:-MYBR3by64MlBF3383MZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/30f067991444dc9364f15c4fba698d08d3a0eaa4/tools/dynamic_analysis/dump_limit_malloc_stacks#L2)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

FWIW, you can always externally kick on debugging of bash scripts. For script `foo`, you can invoke it via `bash -vx foo` and get plenty of diagnostics. If it were invoked by CI, then I'd be in favor of internal `-x`.
</blockquote></details>

Done.

**Comment by rpoyner-tri** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYAq4KK84A-rcB6FYSd:-MYBRcUOAutI_Rtb9tPQ:b-n3zjqw) (no related file):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

@drake-jenkins-bot mac-catalina-clang-bazel-experimental-debug please.
</blockquote></details>

@drake-jenkins-bot mac-catalina-clang-bazel-experimental-debug please.

---

**Review by EricCousineau-TRI — COMMENTED** [review]
> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14899)* status: 3 unresolved discussions, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers (waiting on @EricCousineau-TRI and @rpoyner-tri)

___
*[common/test_utilities/limit_malloc.cc, line 175 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYB456H3A_daP3zExuM:-MYBcnQQElZIOtllEW1q:b69p3ay) ([raw file](https://github.com/robotlocomotion/drake/blob/30f067991444dc9364f15c4fba698d08d3a0eaa4/common/test_utilities/limit_malloc.cc#L175)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

I don't think I can trust `std::string` not to allocate, so things will be a bit more primitive...
</blockquote></details>

OK Ah, good point!
___
*[common/test_utilities/test/limit_malloc_manual_test.cc, line 5 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYBcvXsD3i8JRJK5XYl:-MYBcvXt9CiCQ9l5Idqj:b86viim) ([raw file](https://github.com/robotlocomotion/drake/blob/43ef7a5fe32593d340d3ab717d64d543215990c2/common/test_utilities/test/limit_malloc_manual_test.cc#L5)):*
> ```cpp
> int main(void) {
>   drake::test::LimitMalloc guard;
>   new int;  // Deliberately cause LimitMalloc to report failure.
> ```

nit Unclear why you introduce an extra defect, leaked memory. (Naively, I could think that's part of the purpose of this test.)

Can you instead use something like `std::make_unique<int>()`, or just comment that leaking memory is not the point of this test?
___
*[tools/dynamic_analysis/dump_limit_malloc_stacks, line 1 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYB4sxM7_hoHC_IlnNZ:-MYBd96L6THY3wcCGnYU:bh226ey) ([raw file](https://github.com/robotlocomotion/drake/blob/30f067991444dc9364f15c4fba698d08d3a0eaa4/tools/dynamic_analysis/dump_limit_malloc_stacks#L1)):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

I'm inclined to punt on another level of automated test at this point.
</blockquote></details>

OK IMO not that hard to try, but def. won't force the issue!
___
*[tools/dynamic_analysis/dump_limit_malloc_stacks, line 2 at r1](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYB3t0h4f9aMnrjQrrc:-MYBdKIZC1i9OGd5OE6b:br4j0dv) ([raw file](https://github.com/robotlocomotion/drake/blob/30f067991444dc9364f15c4fba698d08d3a0eaa4/tools/dynamic_analysis/dump_limit_malloc_stacks#L2)):*
> FWIW, you can always externally [...]

OK Noted - thanks for the info!
___
*[tools/dynamic_analysis/dump_limit_malloc_stacks, line 11 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYBdRudAOSxut6_-bC9:-MYBdRudAOSxut6_-bCA:b-pp4wlz) ([raw file](https://github.com/robotlocomotion/drake/blob/43ef7a5fe32593d340d3ab717d64d543215990c2/tools/dynamic_analysis/dump_limit_malloc_stacks#L11)):*
> ```
> # Example:
> #
> #  $ bazel build -c dbg //common/test_utilities:limit_malloc_manual_test
> ```

Working: Running locally.

**Review by EricCousineau-TRI — COMMENTED** [review]
:lgtm: - thanks!

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14899#-:-MYBe6i12XDd-L7rJoxA:b-hidk3u)* status: 3 unresolved discussions, needs at least two assigned reviewers (waiting on @rpoyner-tri)

___
*[tools/dynamic_analysis/dump_limit_malloc_stacks, line 11 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYBdRudAOSxut6_-bC9:-MYBe3HqE0URj9msIpFW:b-r8r3rk) ([raw file](https://github.com/robotlocomotion/drake/blob/43ef7a5fe32593d340d3ab717d64d543215990c2/tools/dynamic_analysis/dump_limit_malloc_stacks#L11)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Working: Running locally.
</blockquote></details>

OK Sweet! w/ fixed typo, confirmed works as intended.
___
*[tools/dynamic_analysis/dump_limit_malloc_stacks, line 13 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYBdu0_-xJL2FKY0MJm:-MYBdu0_-xJL2FKY0MJn:b-7j91jm) ([raw file](https://github.com/robotlocomotion/drake/blob/43ef7a5fe32593d340d3ab717d64d543215990c2/tools/dynamic_analysis/dump_limit_malloc_stacks#L13)):*
> ```
> #  $ bazel build -c dbg //common/test_utilities:limit_malloc_manual_test
> #  $ tools/dynamic_analysis/dump_limit_malloc_stacks \
> #      bazel-bin/common_test_utilities/limit_malloc_manual_test \
> ```

Typo: Should be `common/test_utilities`, not `common_test_utilities`

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14899)* status: 2 unresolved discussions, needs at least two assigned reviewers (waiting on @EricCousineau-TRI and @rpoyner-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYAq4KK84A-rcB6FYSd:-MYBjkOf2vly0hoAV-OI:bhl2802) (no related file):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

@drake-jenkins-bot mac-catalina-clang-bazel-experimental-debug please.
</blockquote></details>

@drake-jenkins-bot mac-catalina-clang-bazel-experimental-debug please.

---
___
*[tools/dynamic_analysis/dump_limit_malloc_stacks, line 13 at r2](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYBdu0_-xJL2FKY0MJm:-MYBjgSb5VQOQEhR_ED3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/43ef7a5fe32593d340d3ab717d64d543215990c2/tools/dynamic_analysis/dump_limit_malloc_stacks#L13)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Typo: Should be `common/test_utilities`, not `common_test_utilities`
</blockquote></details>

Done.

**Review by EricCousineau-TRI — COMMENTED** [review]
> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14899)* status: 1 unresolved discussion, needs at least two assigned reviewers (waiting on @rpoyner-tri)

**Review by rpoyner-tri — COMMENTED** [review]
+@soonho-tri for platform review; Wednesday schedule. Hopefully the mac build will be done by then. :)

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14899#-:-MYBnNVJB8EhXc8Om4mg:b-s9in6s)* status: 1 unresolved discussion, LGTM missing from assignee soonho-tri(platform) (waiting on @rpoyner-tri and @soonho-tri)

**Comment by rpoyner-tri** [comment]
*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYAq4KK84A-rcB6FYSd:-MYBsXBFCmqZoUyes9hI:bhl2802) (no related file):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

@drake-jenkins-bot mac-catalina-clang-bazel-experimental-debug please.
</blockquote></details>

@drake-jenkins-bot mac-catalina-clang-bazel-experimental-debug please.

---

**Review by soonho-tri — COMMENTED** [review]
:lgtm:

> Reviewed 2 of 4 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14899#-:-MYF9R96E8NTldHJfhfS:bnfp4nl)* status: 1 unresolved discussion (waiting on @rpoyner-tri)

**Review by rpoyner-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/14899)* status: :shipit: complete! all discussions resolved, LGTM from assignees soonho-tri(platform),EricCousineau-TRI(platform) (waiting on @rpoyner-tri)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/14899#-MYAq4KK84A-rcB6FYSd:-MYFwZy65hKRRuYYaApn:bmfs3rt) (no related file):*
<details><summary><i>Previously, rpoyner-tri (Rick Poyner (rico)) wrote…</i></summary><blockquote>

@drake-jenkins-bot mac-catalina-clang-bazel-experimental-debug please.
</blockquote></details>

passed.

---

**Pull request merged**